pkg/expectationsmgr/expectationsmgr.go

```diff
@@ -0,0 +1,58 @@
+package expectationsmgr
+
+import (
+	"context"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+)
+
+const tickDuration = 3 * time.Second
+
+// ExpectationsMgr fires after data expectations have been populated in the ready Tracker and runs
+// until the Tracker is satisfeid or the process is exiting. It removes Data expectations for any
+// GVKs that are expected in the Tracker but not watched by the CacheManager.
+type ExpectationsMgr struct {
+	cacheMgr *cachemanager.CacheManager
+	tracker  *readiness.Tracker
+}
+
+func NewExpecationsManager(cm *cachemanager.CacheManager, rt *readiness.Tracker) *ExpectationsMgr {
+	return &ExpectationsMgr{
+		cacheMgr: cm,
+		tracker:  rt,
+	}
+}
+
+func (e *ExpectationsMgr) Run(ctx context.Context) {
+	ticker := time.NewTicker(tickDuration)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if e.tracker.Satisfied() {
+				// we're done, there's no need to
+				// further manage the data sync expectations.
+				return
+			}
+			if !(e.tracker.DataPopulated() && e.cacheMgr.Started()) {
```

Comment:
Don't we want to wait on the `cachemaanger` to signal that it has been started by the controller too?

Otherwise, the `expectationsmgr/ pruner` could race with the `cachemanger`'s Upsert/RemoveSource calls where we remove data gvks too soon, right? 

Since the `cachemanager` is a "runnable", it's started last by the controller so, technically, we have some delay between the  config and sync controllers firing off to add to the `cachemanager`'s watched gvks. Wdyt?

Reply:
This is moot. Whether cache manager is running says nothing as to whether it's up-to-date. Whether cache manager is up-to-date is based on whether syncsets/configs have been ingested.

Cache manager does not need to be running for its internal state to be current. Those updates happen in the syncset/config controller threads. The runnable starts background tasks that clean stale state, which is irrelevant to expectations.

Reply:
i.e. even if cache manager running was necessary, it would still not be a sufficient condition to say "now it's time to rely on me"